### DISCUSSION TITLE
Wander_3D

### DISCUSSION THREAD

**Discussion by hana-gharrad** [gama-platform/gama#55] [Q&A]
Hello all, 
I'm trying to use the wander in 3D moves (Gama 1.9.1). However, I got wander_3D does not exist!
Am I missing something? Here is the code:
`species agent2 skills: [moving, moving3D] control:simple_bdi {
	float view_dist<-150.0;
	float speed <- 2#km/#h;
	rgb my_color<-rnd_color(255);
	init {
		do add_belief(ready);
	}
	aspect basic {
	  draw circle(30) color: my_color border: #black;
	  draw circle(view_dist) color: my_color border: #black wireframe: true;
	  draw name at:location color:my_color;
	}
	rule belief: ready new_desire: fly strength: 3.0;
	plan do_fly intention: fly{
	do wander_3D;
	} 
} `
Referring to the documentation. I need just 'moving' in the skills but I added both. The agents does not move in 3d, it moves only in 2d (when I change to wander). 

I appreciate your help
Kind regards,

**Comment by agrignard** [comment]
Hi @hana-gharrad that's actually a good question. When the 2 skills are used simultaneously it seems that the wander used by defulat is the wander in 2D. What about using only moving3D skills in your model?

**Reply by hana-gharrad** [reply]
Using only moving3D as skill. The error message says: The variable wander_3D is not defined or accessible in this context. If I use wander with moving3D as skill. The agent does not move.

**Reply by agrignard** [reply]
If I am not wrong ```wander_3D``` doesn't exist anymore indeed. The idea is that if you use the ```moving```skills wander will be in 2D if you use the ```moving3D``` skills the wander will be in 3D.

Is it possible for you to share the full model so I can test it?

**Comment by hana-gharrad** [comment]
Sure, here is the code. Thank you. So the z_max will not be applicable anymore? 
```gaml
/**
* Name: Agent
* Based on the internal empty template. 
* Author: hana
* Tags: 
*/


model AgentBDI

global {
		string in_1_1 <- "in_1_1";
		predicate sick<- new_predicate("sick") ;
		predicate walk_slowly_intention<- new_predicate("walk_slowly") ;
		
		predicate hungry<- new_predicate("hungry") ;
		predicate take_break_eat_intention<- new_predicate("take_break_eat_intention") ;
		predicate go_upper <- new_predicate("Go uppper") ;
		geometry shape <- square(20 #km);
		init {
			create agent2 number:5;
		}
		}
		



grid my_grid width: 2 height: 2 {
	init {
			color <- rgb(0,0,0, 0.5);
		
	}
}

species agent2 skills: [moving3D] control:simple_bdi {
    
	float view_dist<-1000.0;
	my_grid cell <- nil;
	float speed <- 2#km/#h;
	rgb my_color<-rnd_color(255);

	init {
		cell <- one_of(my_grid where (each.shape overlaps self.shape));

	write (cell);
	write (obedience);
	do add_belief(sick);
	}
	action eat {
		int cycle_now <- cycle;
		if cycle=cycle_now+300{
			do remove_belief(hungry);
		}
	}
	rule belief: sick new_desire: walk_slowly_intention strength: 3.0;
	rule belief: hungry new_desire: take_break_eat_intention strength: 4.0;
	
	aspect default {
	  draw circle(200) color: my_color border: #black;
	  draw circle(view_dist) color: my_color border: #black wireframe: true;
	}
	perceive target:self {
		 my_grid m <- one_of(my_grid where (each.shape overlaps self.shape));
		if( m!=self.cell) {
			if (m.grid_x = 0 and m.grid_y=0) {
							do add_belief(hungry);
							do remove_desire(walk_slowly_intention);
							do remove_desire(take_break_eat_intention);
							
				
			}
		self.cell<-m;
		}
	}

	law working belief: hungry new_obligation: go_upper threshold:0.8;

	norm share_information obligation: go_upper threshold: 0.7 finished_when: has_belief(go_upper) instantaneous: true {
	do add_belief(go_upper);
	}

	plan do_wander intention: walk_slowly_intention{
		do wander amplitude: 5.5;
	}
	plan do_take_break intention: take_break_eat_intention{
	
		do eat;
	}

}
    experiment GoldBdi type: gui {
	output {
		display map type: 3d {
		grid my_grid lines: #black ;
		species agent2;
		}
	}
	
	}

```

**Comment by lesquoyb** [comment]
Hello, I know it's old but just if someone is having the same problems. Here the issue was that your world is a 2d shape and thus 3d moving is not possible, you can fix that by creating a cube instead of a square.
Closing the discussion as everything was answered
